File: matlab/atomistic.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%        PROGRAM FOR RESIDUAL INTERACTION PROBABILITY CALCULATION
%               USING ATOMISTIC VAN DER WAALS POTENTIALS
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Input files:
% - vdw_cns.param, parameter file used in MD (for obtaining atom types)
% - pdb_cns.top, topology file used in MD (for obtaining energy values)
% Output of the program:
% - normalized energy matrix
% - total energy matrix
% weight: e^-E/kt
% normalized: normalized version of totalenergies matrix

function y = atomistic(ProteinName)
File = 'vdw_cns.param';
rcut = 5; % cutoff radius (Å)

fid = fopen(File);
i = 0;
while feof(fid) == 0
    tline = fgets(fid,255);
    if  tline(2:10) == 'NONBonded'
        i = i+1;
        partype(i,:)  = tline(13:16);
        pareps(i,1)   = str2num(tline(21:26));
        parsigma(i,1) = str2num(tline(30:35));
    end
end
fclose(fid);

File = 'pdb_cns.top';
fid = fopen(File);
i = 0;
while feof(fid) == 0
    tline = fgets(fid,255);
    i = i+1;
    parrestype(i,:) = tline(1:3);
    paratomtype(i,:) = tline(11:13);
    partypecor(i,:) = tline(21:24); % corresponding type to atomtype
    charge(i,:) = tline(34:38);
end
fclose(fid); 

kT = 1;
pdb = pdbreader(ProteinName);
k = 1;
mult_count = 0;
for i = 1:length(pdb.Model.Atom)
    if strcmp(pdb.Model.Atom(1,i).iCode,'') & strcmp(pdb.Model.Atom(1,i).altLoc,'')
        atomnumber1(k,1) = pdb.Model.Atom(1,i).AtomSerNo;
        atomtype1{k}     = upper(pdb.Model.Atom(1,i).AtomName);
        altloc{k} = [];
        if length(pdb.Model.Atom(1,i).altLoc) > 0
            altloc{k} = pdb.Model.Atom(1,i).altLoc;
        end
        restype1{k}     = upper(pdb.Model.Atom(1,i).resName);
        chainid1{k}     = upper(pdb.Model.Atom(1,i).chainID);
        resnumber1(k,1) = pdb.Model.Atom(1,i).resSeq;
        x1(k,1)         = pdb.Model.Atom(1,i).X;
        y1(k,1)         = pdb.Model.Atom(1,i).Y;
        z1(k,1)         = pdb.Model.Atom(1,i).Z;
        bfactor1(k,1)   = 1;
        rfactor1(k,1)   = 1;
        element1{i}     = pdb.Model.Atom(1,i).element;
        k = k+1;
    else
        mult_count = mult_count+1;
    end
end
ds = length(pdb.Model.Atom) - mult_count;

CorFile = strcat(ProteinName, '.cor');
rcor = dlmread(CorFile); % input obtained from readpdb.f

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Filter hydrogens/altLocs as original logic
sum  = 0; sumx = 0; sumy = 0; sumz = 0; sumrf = 0;
j = 0; k = -1; i = 0;
while i ~= ds
    i = i+1;
    j = j+1;
    if strcmp(restype1{i},'HIP')==1 || strcmp(restype1{i},'HIE')==1 || strcmp(restype1{i},'HID')==1 || strcmp(restype1{i},'HSD')==1 || strcmp(restype1{i},'HSE')==1 || strcmp(restype1{i},'HSP')==1 
        restype1{i} = 'HIS';
    elseif strcmp(restype1{i},'LYN')==1
        restype1{i} = 'LYS';
    elseif strcmp(restype1{i},'GLH')==1 
        restype1{i} = 'GLU';
    elseif strcmp(restype1{i},'CYM')==1 || strcmp(restype1{i},'CYX')==1
        restype1{i} = 'CYS';
    elseif strcmp(restype1{i},'ASH')==1
        restype1{i} = 'ASN';
    elseif strcmp(restype1{i},'TYM')==1
        restype1{i} = 'TYR';
    end
    if i ~= ds
        while strcmp(element1{i},'H')==1 || strcmp(atomtype1{i},'H')==1 || strcmp(atomtype1{i},'D')==1 || strcmp(atomtype1{i},'E')==1 || strcmp(atomtype1{i},'G')==1 || strcmp(atomtype1{i},'HN')==1 || (any(atomtype1{i}=='H')==1 & ~any(atomtype1{i}=='N'))
            i = i+1; % skip hydrogens
            if ds == i, break; end
        end
    end
    if i==ds && ( strcmp(element1{i},'H')==1 || strcmp(atomtype1{i},'H')==1 || strcmp(atomtype1{i},'D')==1 || strcmp(atomtype1{i},'E')==1 || strcmp(atomtype1{i},'G')==1 || strcmp(atomtype1{i},'HN')==1 || (any(atomtype1{i}=='H')==1 & ~any(atomtype1{i}=='N')) )
        j = j-1;
        break
    end
    if length(altloc{i})~=0 & length(altloc{i+1})~=0 & strcmp(altloc{i},altloc{i+1})==0 
        j = j-1;
        continue;
    end
    atomnumber(j,1) = atomnumber1(i,1);
    atomtype{j}     = atomtype1{i};
    restype{j}      = restype1{i};
    chainid{j}      = chainid1{i};
    resnumber(j,1)  = resnumber1(i,1);
    x(j,1)          = x1(i,1);
    y(j,1)          = y1(i,1);
    z(j,1)          = z1(i,1);
    bfactor(j,1)    = bfactor1(i,1);
    rfactor(j,1)    = rfactor1(i,1);
    element{j}      = element1{i};
end
ds = j;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Map LJ params
for i = 1:ds
    if strcmp(atomtype{i},'OXT')==1 || strcmp(atomtype{i},'O2')==1 || strcmp(atomtype{i},'O1')==1 || strcmp(element{i},'O')==1
        atomtype{i} = 'O';
    end
    for j = 1:247
        ra = safe_strtrim(parrestype(j,:));
        pa = safe_strtrim(paratomtype(j,:));
        if strcmp(restype{i}, ra)==1 & strcmp(pa, atomtype{i})==1
            for k = 1:34
                if strcmp(char(partypecor(j,:)), char(partype(k,:)))==1
                    eps(i,1)   = pareps(k,1);
                    sigma(i,1) = parsigma(k,1);
                end
            end
        end
    end
end

% Keep only atoms that received parameters (original code already resets ds)
ds = size(eps,1);

fprintf('\n\nReading epsilon and sigma values are over\n\n');
w = 0;
for i = 1:ds
    if eps(i,1)==0 || sigma(i,1)==0
        w = w+1;
        error(w,1) = atomnumber(i,1);
        berror{w}  = restype{i};
    end
end
if w > 0
    fprintf('\n\nThere is an error\n\n');
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% ENERGY CALCULATION (LJ 12-6)
for i = 1:ds
    for j = 1:ds
        r(i,j) = sqrt((x(i,1)-x(j,1))^2 + (y(i,1)-y(j,1))^2 + (z(i,1)-z(j,1))^2);
        sigmamix(i,j) = (sigma(i,1)+sigma(j,1))/2;
        epsmix(i,j)   = sqrt((eps(i,1)*eps(j,1)));
        rmin(i,j)     = (2^(1/6))*sigmamix(i,j);
        if i==j
            energy(i,j) = 0;
        elseif r(i,j) < rmin(i,j)
            energy(i,j) = 4*epsmix(i,j)*((sigmamix(i,j)/rmin(i,j))^12 - (sigmamix(i,j)/rmin(i,j))^6);
        elseif r(i,j) > rcut
            energy(i,j) = 0;
        else
            energy(i,j) = 4*epsmix(i,j)*((sigmamix(i,j)/r(i,j))^12 - (sigmamix(i,j)/r(i,j))^6);
        end
    end
    if mod(i,1000)==0
        fprintf('Atoms calculated so far: %d\n\n', i);
    end
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% RESIDUAL ENERGY CALCULATION
A = energy;                 % (ds x ds)
[Asz1, Asz2] = size(A);     % should be ds x ds
xsz = size(rcor);
totresno = xsz(1);

resenergy = zeros(totresno, totresno); % preallocate
sum_val = 0;
kalani = 1;
kalanj = 1;

for l = 1:totresno
    for k = 1:totresno
        % Intended atom-block indices from .cor
        i_start = kalani;
        i_end   = kalani + rcor(l,9) - 1;
        j_start = kalanj;
        j_end   = kalanj + rcor(k,9) - 1;

        % ---- Bounds clamp to available matrix A (minimal change, avoids OOB)
        if i_start > Asz1 || j_start > Asz2
            sum_val = 0; % nothing to add if block starts beyond A
        else
            i_end_eff = min(i_end, Asz1);
            j_end_eff = min(j_end, Asz2);
            sum_val = 0;
            for ii = i_start:i_end_eff
                for jj = j_start:j_end_eff
                    sum_val = sum_val + A(ii,jj);
                end
            end
        end

        resenergy(l,k) = sum_val/2;
        if l == k
            resenergy(l,k) = 0;
        end

        if k ~= totresno
            kalanj = kalanj + rcor(k,9);
        end
    end
    kalanj = 1;
    kalani = kalani + rcor(l,9);
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% ENERGY NORMALIZATION
sum_row = 0;
for i = 1:totresno
    for j = 1:totresno
        weight(i,j) = exp(-resenergy(i,j)/kT);  % weight = exp(-E/kt)
        sum_row = weight(i,j) + sum_row;        % row sum
    end
    for j = 1:totresno
        normalized(i,j) = weight(i,j)/sum_row;  % normalization
        if i == j
            normalized(i,j) = 0;
        end
    end
    sum_row = 0;
end

File = strcat(ProteinName, '_atomistic.out');
dlmwrite(File, normalized, '\t'); % probability matrix
y = 0; % keep original function signature
end

% ------------------------- Helper (Octave-safe) --------------------------
function s = safe_strtrim(x)
% Accept char, string, or numeric arrays and return a trimmed char row.
  if ischar(x)
    s = strtrim(x);
  elseif isstring(x)
    s = strtrim(char(x));
  elseif isnumeric(x)
    s = strtrim(char(x));
  else
    s = strtrim(char(x));
  end
end

File: mcpath/atomistic.py
# -*- coding: utf-8 -*-
"""
PROGRAM FOR RESIDUAL INTERACTION PROBABILITY CALCULATION
USING ATOMISTIC VAN DER WAALS POTENTIALS (Python / Colab version)

Inputs (same as MATLAB version):
- 'vdw_cns.param' : parameter file (for atom-type -> epsilon,sigma)
- 'pdb_cns.top'   : topology file  (for (res,atom) -> parameter-type mapping)
- '<ProteinName>.pdb' : the structure file
- '<ProteinName>.cor' : coordinates/atom-block info (from your readpdb workflow)

Output:
- '<ProteinName>_atomistic.out' : normalized probability matrix (tab-separated)

Notes:
- kT is set to 1.0 to reproduce your MATLAB behavior exactly
- Hydrogens & altLoc handling are matched to your MATLAB rules
- For r < rmin, energy uses the constant value with rmin in the denominator (same as your code)
- For r > rcut, energy = 0
- Diagonal energies forced to 0; diagonal of normalized matrix forced to 0
"""

import os
import re
import math
import numpy as np

# ---- Install Biopython if missing (for clean PDB parsing) ----
try:
    import Bio  # noqa: F401
except ImportError:
    import sys, subprocess
    subprocess.check_call([sys.executable, "-m", "pip", "install", "biopython"])
from Bio.PDB import PDBParser


# -------------------------- Helpers --------------------------

def _trim(s):
    """Trim and return a plain Python str even if the input is bytes/NumPy scalar."""
    if isinstance(s, bytes):
        s = s.decode("utf-8", "ignore")
    return str(s).strip()


def parse_vdw_params(param_path):
    """
    Parse 'vdw_cns.param' lines that start with ' NONBonded' (1-based in MATLAB).
    We expect fixed columns similar to the MATLAB slicing:
      partype  = cols 13-16
      pareps   = cols 21-26
      parsigma = cols 30-35
    Returns: dict[partype] = (eps, sigma)
    """
    partype_to_epssigma = {}
    with open(param_path, "r", encoding="utf-8", errors="ignore") as f:
        for line in f:
            # MATLAB used tline(2:10) == 'NONBonded' (1-based),
            # which equals Python line[1:10]
            if len(line) >= 10 and line[1:10] == "NONBonded":
                seg = line.rstrip("\n")

                def field(seg, i, j):
                    # Python slices are 0-based, end-exclusive
                    return _trim(seg[i:j]) if len(seg) >= j else _trim(seg[i:])

                partype  = field(seg, 12, 16)  # 13:16 in MATLAB
                pareps_s = field(seg, 20, 26)  # 21:26
                parsig_s = field(seg, 29, 35)  # 30:35

                if partype:
                    try:
                        eps   = float(pareps_s)
                        sigma = float(parsig_s)
                        partype_to_epssigma[partype] = (eps, sigma)
                    except ValueError:
                        # tolerate malformed numerics like MATLAB script
                        pass
    return partype_to_epssigma


def parse_topology(top_path):
    """
    Parse 'pdb_cns.top' with MATLAB columns:
      parrestype = cols 1:3
      paratomtype= cols 11:13
      partypecor = cols 21:24  # (map to vdw param 'partype')
      charge     = cols 34:38  # (unused here)
    Returns: dict[(res, atom)] -> mapped_type
    """
    mapping = {}
    with open(top_path, "r", encoding="utf-8", errors="ignore") as f:
        for line in f:
            seg = line.rstrip("\n")

            def field(seg, i, j):
                return _trim(seg[i:j]) if len(seg) >= j else _trim(seg[i:])

            res  = field(seg, 0, 3)    # 1:3 -> 0:3
            atm  = field(seg, 10, 13)  # 11:13 -> 10:13
            corr = field(seg, 20, 24)  # 21:24 -> 20:24

            if res and atm and corr:
                mapping[(res.upper(), atm.upper())] = corr
    return mapping


def load_pdb_atoms(pdb_path):
    """
    Load atoms from PDB (like your pdbreader).
    Filters:
      - Only atoms with iCode='' and altLoc in {'', ' ', 'A'} (choose first conformer)
      - Skip hydrogens via (element=='H') or name rules
    Returns dict with arrays for matched heavy atoms.
    """
    parser = PDBParser(QUIET=True)
    structure = parser.get_structure("prot", pdb_path)

    serials, atom_names, resnames, chains, resseqs = [], [], [], [], []
    xs, ys, zs = [], [], []
    elements, altlocs, icodes = [], [], []

    # choose first altLoc seen for each (chain, resseq, icode, atom_name)
    seen_altloc_key = set()

    for model in structure:
        for chain in model:
            for residue in chain:
                het, resseq, icode = residue.id
                resname = _trim(residue.get_resname()).upper()
                for atom in residue:
                    atom_name = _trim(atom.get_name()).upper()
                    element = _trim(atom.element).upper() if atom.element else ""
                    altloc = _trim(getattr(atom, "altloc", "")).upper()
                    serial = int(atom.get_serial_number())
                    key = (chain.id, resseq, icode, atom_name)

                    if key in seen_altloc_key:
                        continue
                    if altloc not in ("", " ", "A"):
                        continue
                    seen_altloc_key.add(key)

                    # hydrogen rules (match MATLAB)
                    if element == "H":
                        continue
                    if atom_name.startswith("H") and ("N" not in atom_name):
                        continue
                    if atom_name in ("D", "E", "G", "HN"):
                        continue

                    # residue normalization
                    resname_norm = resname
                    if resname_norm in ("HIP", "HIE", "HID", "HSD", "HSE", "HSP"):
                        resname_norm = "HIS"
                    elif resname_norm == "LYN":
                        resname_norm = "LYS"
                    elif resname_norm == "GLH":
                        resname_norm = "GLU"
                    elif resname_norm in ("CYM", "CYX"):
                        resname_norm = "CYS"
                    elif resname_norm == "ASH":
                        resname_norm = "ASN"
                    elif resname_norm == "TYM":
                        resname_norm = "TYR"

                    x, y, z = atom.coord
                    serials.append(serial)
                    atom_names.append(atom_name)
                    resnames.append(resname_norm)
                    chains.append(_trim(chain.id).upper())
                    resseqs.append(int(resseq))
                    xs.append(float(x)); ys.append(float(y)); zs.append(float(z))
                    elements.append(element)
                    altlocs.append(altloc)
                    icodes.append(_trim(icode))

    xyz = np.column_stack([xs, ys, zs]).astype(float)
    return {
        "serial": np.array(serials, dtype=int),
        "atom":   atom_names,
        "res":    resnames,
        "chain":  chains,
        "resseq": np.array(resseqs, dtype=int),
        "xyz":    xyz,
        "element": elements,
        "altloc": altlocs,
        "icode":  icodes,
    }


def map_eps_sigma(atoms, top_map, vdw_map):
    """
    Map each atom to (eps, sigma) via:
       (res, atom) --(pdb_cns.top)--> partype --> (vdw_cns.param) --> (eps,sigma)
    Also perform the MATLAB 'O' normalizations for atom names:
       if atom in {'OXT','O2','O1'} or element=='O' => atom = 'O'
    Return filtered atoms (only those with mapped params) and arrays eps, sigma.
    """
    res  = atoms["res"]
    atm  = atoms["atom"]
    elem = atoms["element"]

    # Normalize atom name for O variants
    norm_atom = []
    for a, el in zip(atm, elem):
        norm_atom.append("O" if (a in ("OXT", "O2", "O1") or el == "O") else a)

    eps_list, sig_list, keep_idx = [], [], []
    for i, (rname, aname) in enumerate(zip(res, norm_atom)):
        key   = (rname.upper(), aname.upper())
        ptype = top_map.get(key)
        if ptype is None:
            continue
        es = vdw_map.get(ptype)
        if es is None:
            continue
        eps, sig = es
        if (eps == 0.0) or (sig == 0.0):
            continue
        keep_idx.append(i)
        eps_list.append(float(eps))
        sig_list.append(float(sig))

    if not keep_idx:
        raise RuntimeError("No atoms found with valid (eps, sigma) mapping. "
                           "Check that pdb_cns.top and vdw_cns.param match your PDB.")

    keep_idx = np.array(keep_idx, dtype=int)

    # --- FIXED: build filtered atom dict without the undefined `val` name ---
    filtered_atoms = {}
    for k, v in atoms.items():
        if k == "xyz":
            # handled separately for 2D slice
            continue
        if isinstance(v, np.ndarray):
            filtered_atoms[k] = v[keep_idx]
        elif isinstance(v, list):
            filtered_atoms[k] = [v[j] for j in keep_idx]
        else:
            # metadata scalars/unused fields are copied as-is
            filtered_atoms[k] = v
    # xyz is (N,3)
    filtered_atoms["xyz"] = atoms["xyz"][keep_idx, :]

    return filtered_atoms, np.array(eps_list, dtype=float), np.array(sig_list, dtype=float)


def compute_lj_energy(coords, eps, sigma, rcut=5.0, verbose_every=1000):
    """
    Compute full ds x ds LJ 12-6 energy with:
      - Lorentz-Berthelot mixing: sigma_mix = (sigma_i + sigma_j)/2
                                   eps_mix   = sqrt(eps_i * eps_j)
      - rmin = 2^(1/6) * sigma_mix
      - if i==j -> 0
      - if r < rmin -> 4*eps_mix*((sigma_mix/rmin)^12 - (sigma_mix/rmin)^6)  (constant cap)
      - if r > rcut -> 0
      - else       -> 4*eps_mix*((sigma_mix/r)^12   - (sigma_mix/r)^6)
    """
    xyz = coords.astype(float)
    ds  = xyz.shape[0]

    diff = xyz[:, None, :] - xyz[None, :, :]
    r = np.linalg.norm(diff, axis=2)

    sigma_i = sigma.reshape(-1, 1)
    sigma_j = sigma.reshape(1, -1)
    eps_i   = eps.reshape(-1, 1)
    eps_j   = eps.reshape(1, -1)

    sigma_mix = (sigma_i + sigma_j) / 2.0
    eps_mix   = np.sqrt(eps_i * eps_j)
    rmin      = (2.0 ** (1.0 / 6.0)) * sigma_mix

    energy = np.zeros((ds, ds), dtype=float)
    eye_mask = np.eye(ds, dtype=bool)

    with np.errstate(divide="ignore", invalid="ignore"):
        within = (r <= rcut) & (~eye_mask)
        lt_rmin = within & (r < rmin)
        ge_rmin = within & (r >= rmin)

        if np.any(lt_rmin):
            ratio_const = (sigma_mix[lt_rmin] / rmin[lt_rmin])
            e_const = 4.0 * eps_mix[lt_rmin] * (ratio_const**12 - ratio_const**6)
            energy[lt_rmin] = e_const

        if np.any(ge_rmin):
            ratio = sigma_mix[ge_rmin] / r[ge_rmin]
            e = 4.0 * eps_mix[ge_rmin] * (ratio**12 - ratio**6)
            energy[ge_rmin] = e

    np.fill_diagonal(energy, 0.0)
    return energy


def block_sum_residual_energy(A, rcor):
    """
    Residue-level energy block sum (MATLAB replica).
    rcor: column 9 (index 8) is atom count per residue.
    """
    Asz1, Asz2 = A.shape
    if rcor.shape[1] < 9:
        raise ValueError(".cor file must have at least 9 columns; column 9 is atom count per residue.")
    counts = rcor[:, 8].astype(int)
    totresno = counts.shape[0]

    resenergy = np.zeros((totresno, totresno), dtype=float)
    kalani = 0
    for l in range(totresno):
        kalanj = 0
        for k in range(totresno):
            i_start = kalani; i_end = kalani + counts[l]
            j_start = kalanj; j_end = kalanj + counts[k]

            i_start = min(max(i_start, 0), Asz1)
            i_end   = min(max(i_end,   0), Asz1)
            j_start = min(max(j_start, 0), Asz2)
            j_end   = min(max(j_end,   0), Asz2)

            s = float(A[i_start:i_end, j_start:j_end].sum()) if (i_end > i_start and j_end > j_start) else 0.0
            resenergy[l, k] = 0.0 if l == k else 0.5 * s

            if k != totresno - 1:
                kalanj += counts[k]
        kalani += counts[l]
    return resenergy


def row_normalize_weights(resenergy, kT=1.0):
    """weight = exp(-E/kT); row-normalize; zero diagonal."""
    W = np.exp(-resenergy / float(kT))
    row_sums = W.sum(axis=1, keepdims=True)
    row_sums[row_sums == 0.0] = 1.0
    N = W / row_sums
    np.fill_diagonal(N, 0.0)
    return N


def atomistic(ProteinName, param_file="vdw_cns.param", top_file="pdb_cns.top",
              rcut=5.0, kT=1.0, save_txt=True):
    """
    Main driver: replicates MATLAB y = atomistic(ProteinName).
    Creates '<ProteinName>_atomistic.out' and returns the normalized matrix (NumPy array).
    """
    vdw_map = parse_vdw_params(param_file)
    top_map = parse_topology(top_file)
    if not vdw_map:
        raise RuntimeError(f"No NONBonded entries parsed from {param_file}.")
    if not top_map:
        raise RuntimeError(f"No (res,atom)->type mappings parsed from {top_file}.")

    pdb_path = ProteinName if ProteinName.lower().endswith(".pdb") else f"{ProteinName}.pdb"
    if not os.path.isfile(pdb_path):
        raise FileNotFoundError(f"PDB not found: {pdb_path}")
    atoms = load_pdb_atoms(pdb_path)

    atoms_f, eps, sigma = map_eps_sigma(atoms, top_map, vdw_map)
    xyz = atoms_f["xyz"]
    if xyz.shape[0] == 0:
        raise RuntimeError("After mapping, zero atoms remain. Check mapping and filters.")

    energy = compute_lj_energy(xyz, eps, sigma, rcut=rcut)

    cor_path = f"{os.path.splitext(pdb_path)[0]}.cor"
    if not os.path.isfile(cor_path):
        raise FileNotFoundError(f".cor file not found: {cor_path}")
    rcor = np.loadtxt(cor_path)
    if rcor.ndim == 1:
        rcor = rcor.reshape(1, -1)

    resenergy  = block_sum_residual_energy(energy, rcor)
    normalized = row_normalize_weights(resenergy, kT=kT)

    if save_txt:
        out_path = f"{os.path.splitext(pdb_path)[0]}_atomistic.out"
        np.savetxt(out_path, normalized, fmt="%.8f", delimiter="\t")

    return normalized

# Example (in Colab):
# norm = atomistic("4CFR", param_file="mcpath/vdw_cns.param", top_file="mcpath/pdb_cns.top", rcut=5.0, kT=1.0)
